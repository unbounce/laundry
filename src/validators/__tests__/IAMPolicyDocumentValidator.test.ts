import { lint } from '../../index';

describe('IAMPolicyDocumentValidator', () => {
  test.each([
    [{ Bool: { 'aws:SecureTransport': true } }],
    [{ StringEquals: { 'aws:RequestTag/foo': ['bar'] } }],
  ])('valid %s', (condition) => {
    const template = JSON.stringify({
      Resources: {
        Role: {
          Type: 'AWS::IAM::Role',
          Properties: {
            AssumeRolePolicyDocument: {},
            Policies: [{
              PolicyName: '',
              PolicyDocument: {
                Version: '',
                Statement: [
                  {
                    Sid: '',
                    Action: '',
                    Effect: 'Allow',
                    Principal: '',
                    Resource: '',
                    Condition: condition
                  }
                ]
              }
            }]
          }
        }
      }
    });
    expect(lint(template)).toEqual([]);
  });
  test('valid, with resource list', () => {
    const template = JSON.stringify({
      Resources: {
        Role: {
          Type: 'AWS::IAM::Role',
          Properties: {
            AssumeRolePolicyDocument: {},
            Policies: [{
              PolicyName: '',
              PolicyDocument: {
                Version: '',
                Statement: [
                  {
                    Action: '',
                    Effect: 'Deny',
                    Principal: '',
                    Resource: [''],
                  }
                ]
              }
            }]
          }
        }
      }
    });
    expect(lint(template)).toEqual([]);
  });

  test('missing required properties', () => {
    const template = JSON.stringify({
      Resources: {
        Role: {
          Type: 'AWS::IAM::Role',
          Properties: {
            AssumeRolePolicyDocument: {},
            Policies: [{
              PolicyName: '',
              PolicyDocument: {
                Version: '',
                Statement: [
                  {}
                ]
              }
            }]
          }
        }
      }
    });
    expect(lint(template)).toMatchSnapshot();
  });

  test('invalid conditional', () => {
    const template = JSON.stringify({
      Resources: {
        Role: {
          Type: 'AWS::IAM::Role',
          Properties: {
            AssumeRolePolicyDocument: {},
            Policies: [{
              PolicyName: '',
              PolicyDocument: {
                Version: '',
                Statement: [
                  {
                    Action: '',
                    Effect: 'Allow',
                    Resource: '',
                    Condition: { Foo: { 'aws:SecureTransport': true } }
                  }
                ]
              }
            }]
          }
        }
      }
    });
    expect(lint(template)).toMatchSnapshot();
  });

  test('invalid condition key', () => {
    const template = JSON.stringify({
      Resources: {
        Role: {
          Type: 'AWS::IAM::Role',
          Properties: {
            AssumeRolePolicyDocument: {},
            Policies: [{
              PolicyName: '',
              PolicyDocument: {
                Version: '',
                Statement: [
                  {
                    Action: '',
                    Effect: 'Allow',
                    Resource: '',
                    Condition: { Bool: { 'cats': 'dogs' } }
                  }
                ]
              }
            }]
          }
        }
      }
    });
    expect(lint(template)).toMatchSnapshot();
  });

  test('invalid types', () => {
    const template = JSON.stringify({
      Resources: {
        Role: {
          Type: 'AWS::IAM::Role',
          Properties: {
            AssumeRolePolicyDocument: {},
            Policies: [{
              PolicyName: '',
              PolicyDocument: {
                Version: '',
                Statement: [
                  {
                    Sid: [],
                    Action: [],
                    Effect: [],
                    Principal: [],
                    Resource: {},
                    Condition: '',
                  }
                ]
              }
            }]
          }
        }
      }
    });
    expect(lint(template)).toMatchSnapshot();
  });

  test('invalid effect', () => {
    const template = JSON.stringify({
      Resources: {
        Role: {
          Type: 'AWS::IAM::Role',
          Properties: {
            AssumeRolePolicyDocument: {},
            Policies: [{
              PolicyName: '',
              PolicyDocument: {
                Version: '',
                Statement: [
                  {
                    Action: '',
                    Effect: 'foo',
                    Resource: '',
                  }
                ]
              }
            }]
          }
        }
      }
    });
    expect(lint(template)).toMatchSnapshot();
  });

});
