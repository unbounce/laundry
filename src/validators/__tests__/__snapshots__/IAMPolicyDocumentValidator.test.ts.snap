// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IAMPolicyDocumentValidator invalid condition key 1`] = `
Array [
  Object {
    "message": "key must be one of aws:CurrentTime, aws:EpochTime, aws:MultiFactorAuthAge, aws:MultiFactorAuthPresent, aws:SecureTransport, aws:UserAgent, aws:PrincipalOrgID, aws:PrincipalTag, aws:PrincipalType, aws:Referer, aws:RequestedRegion, aws:RequestTag, aws:SourceAccount, aws:SourceArn, aws:SourceIp, aws:SourceVpc, aws:SourceVpce, aws:TagKeys, aws:TokenIssueTime, aws:userid, aws:username, s:, got cats",
    "path": Array [
      "Root",
      "Resources",
      "Role",
      "Properties",
      "Policies",
      "0",
      "PolicyDocument",
      "Statement",
      "0",
      "Condition",
      "Bool",
      "cats",
    ],
  },
]
`;

exports[`IAMPolicyDocumentValidator invalid conditional 1`] = `
Array [
  Object {
    "message": "key must be one of StringEquals, StringNotEquals, StringEqualsIgnoreCase, StringNotEqualsIgnoreCase, StringLike, StringNotLike, NumericEquals, NumericNotEquals, NumericLessThan, NumericLessThanEquals, NumericGreaterThan, NumericGreaterThanEquals, DateEquals, DateNotEquals, DateLessThan, DateLessThanEquals, DateGreaterThan, DateGreaterThanEquals, Bool, Null, BinaryEquals, IpAddress, NotIpAddress, ArnEquals, ArnLike, ArnNotEquals, ArnNotLike, got Foo, did you mean Bool?",
    "path": Array [
      "Root",
      "Resources",
      "Role",
      "Properties",
      "Policies",
      "0",
      "PolicyDocument",
      "Statement",
      "0",
      "Condition",
      "Foo",
    ],
  },
]
`;

exports[`IAMPolicyDocumentValidator invalid effect 1`] = `
Array [
  Object {
    "message": "must be one of Allow, Deny, got \\"foo\\"",
    "path": Array [
      "Root",
      "Resources",
      "Role",
      "Properties",
      "Policies",
      "0",
      "PolicyDocument",
      "Statement",
      "0",
      "Effect",
    ],
    "value": "\\"foo\\"",
  },
]
`;

exports[`IAMPolicyDocumentValidator invalid types 1`] = `
Array [
  Object {
    "message": "must be a String, got []",
    "path": Array [
      "Root",
      "Resources",
      "Role",
      "Properties",
      "Policies",
      "0",
      "PolicyDocument",
      "Statement",
      "0",
      "Sid",
    ],
    "value": "[]",
  },
  Object {
    "message": "must be a String, got []",
    "path": Array [
      "Root",
      "Resources",
      "Role",
      "Properties",
      "Policies",
      "0",
      "PolicyDocument",
      "Statement",
      "0",
      "Effect",
    ],
    "value": "[]",
  },
  Object {
    "message": "must be a String or must be an Object, got []",
    "path": Array [
      "Root",
      "Resources",
      "Role",
      "Properties",
      "Policies",
      "0",
      "PolicyDocument",
      "Statement",
      "0",
      "Principal",
    ],
    "value": "[]",
  },
  Object {
    "message": "must be a String or must be a List, got {}",
    "path": Array [
      "Root",
      "Resources",
      "Role",
      "Properties",
      "Policies",
      "0",
      "PolicyDocument",
      "Statement",
      "0",
      "Resource",
    ],
    "value": "{}",
  },
  Object {
    "message": "must be an Object, got \\"\\"",
    "path": Array [
      "Root",
      "Resources",
      "Role",
      "Properties",
      "Policies",
      "0",
      "PolicyDocument",
      "Statement",
      "0",
      "Condition",
    ],
    "value": "\\"\\"",
  },
]
`;

exports[`IAMPolicyDocumentValidator missing required properties 1`] = `
Array [
  Object {
    "message": "is required",
    "path": Array [
      "Root",
      "Resources",
      "Role",
      "Properties",
      "Policies",
      "0",
      "PolicyDocument",
      "Statement",
      "0",
      "Effect",
    ],
  },
]
`;
